plugins/parsers/csv/parser.go

```diff
@@ -16,34 +19,111 @@ import (
 
 type TimeFunc func() time.Time
 
+type Metadata struct {
+	key   string
+	value string
+}
+
+type MetadataRegexPattern struct {
+	Re         *regexp.Regexp
+	KeyIndex   int
+	ValueIndex int
+}
+
 type Config struct {
-	ColumnNames       []string `toml:"csv_column_names"`
-	ColumnTypes       []string `toml:"csv_column_types"`
-	Comment           string   `toml:"csv_comment"`
-	Delimiter         string   `toml:"csv_delimiter"`
-	HeaderRowCount    int      `toml:"csv_header_row_count"`
-	MeasurementColumn string   `toml:"csv_measurement_column"`
-	MetricName        string   `toml:"metric_name"`
-	SkipColumns       int      `toml:"csv_skip_columns"`
-	SkipRows          int      `toml:"csv_skip_rows"`
-	TagColumns        []string `toml:"csv_tag_columns"`
-	TimestampColumn   string   `toml:"csv_timestamp_column"`
-	TimestampFormat   string   `toml:"csv_timestamp_format"`
-	Timezone          string   `toml:"csv_timezone"`
-	TrimSpace         bool     `toml:"csv_trim_space"`
-	SkipValues        []string `toml:"csv_skip_values"`
+	ColumnNames        []string `toml:"csv_column_names"`
+	ColumnTypes        []string `toml:"csv_column_types"`
+	Comment            string   `toml:"csv_comment"`
+	Delimiter          string   `toml:"csv_delimiter"`
+	HeaderRowCount     int      `toml:"csv_header_row_count"`
+	MeasurementColumn  string   `toml:"csv_measurement_column"`
+	MetricName         string   `toml:"metric_name"`
+	SkipColumns        int      `toml:"csv_skip_columns"`
+	SkipRows           int      `toml:"csv_skip_rows"`
+	TagColumns         []string `toml:"csv_tag_columns"`
+	TimestampColumn    string   `toml:"csv_timestamp_column"`
+	TimestampFormat    string   `toml:"csv_timestamp_format"`
+	Timezone           string   `toml:"csv_timezone"`
+	TrimSpace          bool     `toml:"csv_trim_space"`
+	SkipValues         []string `toml:"csv_skip_values"`
+	MetadataRows       int      `toml:"csv_metadata_rows"`
+	MetadataSeparators []string `toml:"cvs_metadata_separators"`
+	MetadataTrimSet    string   `toml:"cvs_metadata_trim_set"`
+
+	MetadataSeparatorList MetadataPattern
 
 	gotColumnNames bool
 
-	TimeFunc    func() time.Time
-	DefaultTags map[string]string
+	TimeFunc      func() time.Time
+	DefaultTags   map[string]string
+	DefaultFields map[string]string
```

Comment:
The map contains the actual metadata, `DefaultFields` confused me.
```suggestion
	Metadata map[string]string
```

Reply:
Just like default tags I had to keep the default fields that we have collected from the metadata section.